python/ray/data/dataset.py

```diff
@@ -696,21 +697,22 @@ def _map_batches_without_batch_size_validation(
     def add_column(
         self,
         col: str,
-        fn: Callable[["pandas.DataFrame"], "pandas.Series"],
+        fn: Callable[["pyarrow.Table"], "pyarrow.Array"],
```

Comment:
fn is user-provided and as such this change is breaking compatibility.
This API is implicitly marked as "stable" with `@PublicAPI(api_group=BT_API_GROUP)`.

@bveeramani Please advise on how we should treat this.

Reply:
> One way around this is converting from/to pandas before and after fn is called.

I think we should do this. If we change the type to `Callable[["pyarrow.Table"], "pyarrow.Array"]`, we'd be making a breaking change.

That said, ensuring that we use the correct extension types for objects and tensors while converting between pandas and Arrow might be non-trivial.

So that we can fix the issue for `select_columns` and `drop_columns` quickly (which AFAIK are used far more than `add_column`), let's keep `add_column` as-is for this PR?

Reply:
Ok. I've now made it so that the standard is pandas but one can also switch to pyarrow when desired since that's what we would prefer.